_posts/2019-08-20-service-objects.markdown

```diff
@@ -0,0 +1,100 @@
+
+This is not something we should do in a model. A model is a representation of a plain ruby object that can be instantiated. In this case we are not going to instantiate any validator model, it's a good thing to avoid instantiating objects whenever it's possible.
+
+So if we can't use a model for this task, should we leave it in the controller? Let's try that:
+
+```ruby
+class PageController < ApplicationController
+
+  def index
+    @show_discount = false
+
+    if params[:discount_code].present?
+      discount = DiscountCode.find_by(discount_code: params[:discount_code])
+
+      if discount.present?
+        @show_discount = discount.created_at > 24.hours.ago
+      else
+        DiscountCode.create(discount_code: params[:discount_code])
+        @show_discount = true
+      end
+
+      if @show_discount
+        client = ExternalApi::Client.new
+        @show_discount = client.validate(discount_code)
+      end
+    end
+  end
+end
+```
+
+The problem with this approach is that we just fattened our controller, adding 13 more lines to it. Also, we can ask ourselves: is it really the responsibility of the controller to check if the discount code is valid or not? I would say this is the kind of thing we should create a service for:
+
+```ruby
+class DiscountValidator
+  def self.validate(discount_code)
+    valid_discount = false
+
+    discount = DiscountCode.find_by(discount_code: discount_code)
+
+    if discount.present?
+      valid_discount = discount.created_at > 24.hours.ago
+    else
+      DiscountCode.create(discount_code: discount_code)
+      valid_discount = true
+    end
+
+    if valid_discount
+      client = ExternalApi::Client.new(api_key: MY_API_KEY)
+      valid_discount = client.validate(discount_code)
+    end
+
+    valid_discount
+  end
+end
+```
+
+Now in our controller we just need to call the new service:
+
+```ruby
+class PageController < ApplicationController
+
+  def index
+    @show_discount = false
+
+    if params[:discount_code].present?
+      @show_discount = DiscountValidator.validate(params[:discount_code])
+    end
+  end
+end
+```
+
+It looks much better, right? With this changes we are refactoring our code so when we look to our controller we know exactly what it does without knowing how the discount validation is done. The code is cleaner and easier to read.
+
+An important notice here is the convention for class and method's name. The name of the service class needs to describes the kind of action that is going to be performed, and with that you realized that we are creating specialized classes that holds a single action in your application. This is also why the method name needs to tell your service to do this action. Other examples of method names could be: `call`, `execute`, `run`.
```

Comment:
```suggestion
An important notice here is the convention for class and method's name. The name of the service class needs to describes the kind of action that is going to be performed, and with that you realized that we are creating specialized classes that hold a single action in your application. This is also why the method name needs to tell your service to do this action. Other examples of method names could be: `call`, `execute`, `run`.
```

Reply:
I'm not sure about this change. What is the correct word @ercohen14 ?